crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,201 @@
+
+```py
+def f(x: str | None):
+    if x is None:
+        ...  # x can be narrowed to `str`
+    else:
+        ...  # x can be narrowed to `None`
+```
+
+All singleton types are also sealed types, closed types and single-value types
+(see below for definitions).
+
+## Sealed types
+
+A sealed type is a type for which it is known
+that there is only a finite and pre-determined set of inhabitants at runtime.
+It follows from this that for every sealed type,
+there is also only a finite and pre-determined set of *subtypes* of that type.
+
+A sealed type in Python has the property
+that its proper subtypes[^2] are all disjunct from one another,
+and the union of all the proper subtypes of a sealed type is exactly equal to the sealed type.
+
+All singleton types are sealed types; however, not all sealed types are singleton types.
+
+Examples of sealed types (other than the singleton types listed above) are:
+
+- `bool`:
+
+    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
+    - The only proper subtypes of `bool` are `Literal[True]`, `Literal[False]`, and `Never`.
+        `Literal[True]`, `Literal[False]` and `Never` are all disjunct types,
+        and their union is exactly equal to `bool`.
+
+- Enums: consider the following enum class:
+
+    ```py
+    from enum import Enum, auto
+
+    class Foo(Enum):
+        X = auto()
+        Y = auto()
+        Z = auto()
+    ```
+
+    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
+        `Foo.X` and `Foo.Y`
+    - The only proper subtypes of `Foo` are `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]`, and `Never`.
+        `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]` and `Never` are all disjunct,
+        and the union `Literal[Foo.X, Foo.Y, Foo.Z] | Never` is exactly equal to the type `Foo`.
+
+Because a sealed type is equivalent to the union of all of its proper subtypes,
+for any given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
+`X & ~A == B | C`. To give a Python example:
+
+```py
+from enum import Enum, auto
+
+class X(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+
+def f(var: X):
+    ...  # the type of `var` here is `X` (which is equivalent to `Literal[X.A, X.B, X.C]`)
+
+    if var is X.A:
+        ...  # var is narrowed to `Literal[X.A]`
+    else:
+        ...  # var is narrowed to `Literal[X.B, X.C]`
+```
+
+## Single-value types
+
+A single-value type is a non-empty type for which it is known
+that all inhabitants of the type are equivalent with respect to their runtime value.
+All singleton types are single-value types, but not all single-value types are singleton types.
+
+Examples of single-value types that are not singleton types
+are `Literal["foo"]`, `Literal[b"foo"]`, `Literal[123456]`, and `tuple[Literal[True], Literal[False]]`.
+All runtime inhabitants of the `Literal["foo"]` type are entirely fungible and equal
```

Comment:
I would also suggest avoiding a description like "entirely fungible", because it is not well-defined. Whether the inhabitants are fungible or not is a matter of perspective and what you're doing with them! If you are doing `is` comparisons, they are not fungible. It's better if we state things precisely in terms of some equivalence relation: rather than "entirely fungible and equal", the type inhabitants are "all in the same equivalence class according to some equivalence relation" (the Python equality relation, in our specific case.)

Reply:
Sad -- there are so few situations in which you get to use the word "fungible"; I really thought this was one of them

Reply:
(While I concede that you're pedantically correct, I _do_ struggle to see what you might be doing with numbers such that one `50000000000000000` object would not be fungible with another `50000000000000000` object 😄)

Reply:
Only things you shouldn't be doing with them, for sure! The intent of the language is absolutely that they should be fungible.